packages/react-components/src/DemocracyLocks.tsx

```diff
@@ -20,50 +20,101 @@ interface Props {
+  return {
+    maxBalance: bnMax(...locks.map(({ balance }) => balance)),
+    sorted: locks
+      .map((info): [DeriveDemocracyLock, BN] => [info, info.unlockAt.gt(bestNumber) ? info.unlockAt.sub(bestNumber) : BN_ZERO])
+      .sort((a, b) => a[0].referendumId.cmp(b[0].referendumId))
+      .sort((a, b) => a[1].cmp(b[1]))
+      .sort((a, b) => a[0].isFinished === b[0].isFinished ? 0 : (a[0].isFinished ? -1 : 1))
+      .reduce((sorted: Entry[], [{ balance, isFinished, referendumId, vote }, blocks]): Entry[] => {
+        const isCountdown = blocks.gt(BN_ZERO);
+        const header = <div>#{referendumId.toString()} {formatBalance(balance, { forceUnit: '-' })} {vote.conviction.toString()}</div>;
+        const prev = sorted.length ? sorted[sorted.length - 1] : null;
+
+        if (!prev || (isCountdown || (isFinished !== prev.isFinished))) {
+          sorted.push({
+            details: (
+              <div className='faded'>
+                {isCountdown
+                  ? (
+                    <BlockToTime
+                      blocks={blocks}
+                      label={`${t<string>('{{blocks}} blocks', { replace: { blocks: formatNumber(blocks) } })}, `}
+                    />
+                  )
+                  : isFinished
+                    ? t<string>('lock expired')
+                    : t<string>('ongoing referendum')
+                }
+              </div>
+            ),
+            headers: [header],
+            isCountdown,
+            isFinished
+          });
+        } else {
+          prev.headers.push(header);
+        }
+
+        return sorted;
+      }, [])
+  };
+}
+
 function DemocracyLocks ({ className = '', value }: Props): React.ReactElement<Props> | null {
   const { t } = useTranslation();
   const { api } = useApi();
   const bestNumber = useCall<BN>(api.derive.chain.bestNumber, []);
   const [trigger] = useState(`${Date.now()}-democracy-locks-${++id}`);
+  const [{ maxBalance, sorted }, setState] = useState<State>({ maxBalance: BN_ZERO, sorted: [] });
 
-  if (!value || !value.length || !bestNumber) {
+  useEffect((): void => {
+    bestNumber && setState(
+      groupLocks(t, bestNumber, value)
+    );
+  }, [bestNumber, t, value]);
+
+  if (!sorted.length) {
     return null;
   }
 
-  const max = bnMax(...value.map(({ balance }) => balance));
-  const mapped = value
-    .map((info): [DeriveDemocracyLock, BN] => [info, info.unlockAt.gt(bestNumber) ? info.unlockAt.sub(bestNumber) : BN_ZERO])
-    .sort((a, b) => a[1].cmp(b[1]));
-
   return (
     <div className={className}>
       <Icon
         icon='clock'
         tooltip={trigger}
       />
-      <FormatBalance value={max} />
+      <FormatBalance value={maxBalance} />
       <Tooltip
-        text={mapped.map(([{ balance, isFinished, referendumId, vote }, blocks], index): React.ReactNode => (
+        text={sorted.map(({ details, headers }, index): React.ReactNode => (
           <div
             className='row'
```

Comment:
I was fighting a little in my solution to get this `row` to have an additional class with `margin-top: 0` to get the grouped element even more compactly grouped, in the case details is `undefined`. But it's a nit.

Reply:
Details is non-optional in the entry.